src/object.d

```diff
@@ -4881,3 +4881,108 @@ void __ArrayDtor(T)(T[] a)
     foreach_reverse (ref T e; a)
         e.__xdtor();
 }
+
+/***********************************************************
+ * Used by `__ArrayCast` to obtain a descriptive error message.
+ * It is a template so it can be used by `__ArrayCast` in -betterC
+ * builds.  It is separate from `__ArrayCast` to minimize code
+ * bloat.
+ * Params:
+ *  fromType = name of the type being cast from
+ *  fromSize = total size in bytes of the array being cast from
+ *  toType   = name of the type being cast o
+ *  toSize   = total size in bytes of the array being cast to
+ * Returns:
+ *  a descriptive error message describing the cast failure
+ */
+private string generateArrayCastErrorMessage()(string fromType, size_t fromSize, string toType, size_t toSize) @trusted
+{
+    import core.internal.string : unsignedToTempString;
+    import core.stdc.stdlib : alloca;
+
+    const(char)[][8] msgComponents =
+    [
+        "Cannot cast `"
+        , fromType
+        , "` to `"
+        , toType
+        , "`; an array of size "
+        , unsignedToTempString(fromSize)
+        , " does not align on an array of size "
+        , unsignedToTempString(toSize)
+    ];
+
+    // convert discontiguous `msgComponents` to contiguous string on the stack
+    size_t length = 0;
+    foreach(m ; msgComponents)
+        length += m.length;
+
+    auto msg = (cast(char*)alloca(length))[0 .. length];
+
+    size_t index = 0;
+    foreach(m ; msgComponents)
+        foreach(c; m)
+            msg[index++] = c;
+
+    return cast(string)msg;
+}
+
+@safe @nogc pure nothrow unittest
+{
+    string msg = generateArrayCastErrorMessage("int[]", 10, "long[]", 8);
+    assert(msg == "Cannot cast `int[]` to `long[]`; an array of size 10 does not align on an array of size 8");
+}
+
+/***********************************************************
+ * The compiler lowers expressions of `cast(TTo[])TFrom[]` to
+ * this implementation.
+ * Params:
+ *  from = the array to reinterpret-cast
+ * Returns:
+ *  `from` reinterpreted as `TTo[]`
+ */
+TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @trusted
```

Comment:
Perhaps use the same order as the `cast` expression.

Reply:
The only rule I know of is to put arguments that needs to be explicitly specified first, otherwise follow the order or arguments. Additionally, in very rare cases, you might have 2 templates argument depending on each others (e.g. via a constraint) in which case the order is self-evident.

In this case, this is never called directly by the user so that order doesn't matter much.
Personally, I find `TFrom, TTo` (as currently specified) to be the clearer alternative.

Reply:
Shouldn't array casting be also `pure` and `@nogc`?

Reply:
Ah, the style policy on template attributes has changed. Good!
https://github.com/dlang/dlang.org/commit/f95cb98ea5f8a8314bf8e41193a10a8180255dc8#diff-2ceb1eaea6947a0a0ad321e635cec2e0
```diff
-    $(LI $(I Templated) functions should $(B not) be annotated with attributes
-         as the compiler can infer them.)
+    $(LI If the template arguments for a $(I templated) function affect whether
+         an attribute is appropriate, then the function should $(B not) be
+         annotated with that attribute so that the compiler can infer it.
+         However, if the attribute is not affected by the template arguments
+         (and thus would always be inferred), then the function should be
+         explicitly annotated with that attribute just like a non-templated
+         function would be.)
```